source/credentials_provider_imds.c

```diff
@@ -782,18 +801,27 @@ static void s_imds_on_stream_complete_fn(struct aws_http_stream *stream, int err
     /* try again, just drop the state from the response to the request state by subtracting one.
      * Don't run the state machine in this callback in this case, let the acquire connection callback handle it. */
     if (error_code == AWS_ERROR_HTTP_CONNECTION_CLOSED) {
-        AWS_LOGF_DEBUG(
-            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
-            "id=%p: Connection was closed, retrying the last request on a new connection.",
-            (void *)imds_user_data->imds_provider);
-        /* roll back to the last request we made, and let it retry. */
-        imds_user_data->query_state -= 1;
+        if (imds_user_data->retry_count++ < IMDS_MAX_RETRIES) {
+            AWS_LOGF_DEBUG(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "id=%p: Connection was closed, retrying the last request on a new connection.",
+                (void *)imds_user_data->imds_provider);
+            /* roll back to the last request we made, and let it retry. */
+            s_imds_state_machine_roll_back_to_request_state(imds_user_data);
+        } else {
+            AWS_LOGF_ERROR(
+                AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+                "id=%p: Connection was closed, retries have been exhausted.",
+                (void *)imds_user_data->imds_provider);
+            /* roll back to the last request we made, and let it retry. */
+            imds_user_data->query_state = AWS_IMDS_QS_UNRECOVERABLE_ERROR;
+        }
     } else {
```

Comment:
are there any other error codes that need a check? or termination?

Reply:
shouldn't be